vault/acl.go

```diff
@@ -47,41 +55,141 @@ func NewACL(policies []*Policy) (*ACL, error) {
+			case existing.capabilitiesBitmap&DenyCapabilityInt > 0:
 				// If we are explicitly denied in the existing capability set,
-				// don't save anything else
+				// don't save anything else. Explicit denial entries do not set
+				// the MFA map, so checking the bitmap is sufficient.
 
 			case pc.CapabilitiesBitmap&DenyCapabilityInt > 0:
-				// If this new policy explicitly denies, only save the deny value
-				tree.Insert(pc.Prefix, DenyCapabilityInt)
+				// If this new policy explicitly denies, only save the deny
+				// value and leave the map at nil
+				tree.Insert(pc.Prefix, aclEntry{
+					capabilitiesBitmap: DenyCapabilityInt,
+				})
 
 			default:
 				// Insert the capabilities in this new policy into the existing
 				// value
-				tree.Insert(pc.Prefix, existing|pc.CapabilitiesBitmap)
+				tree.Insert(pc.Prefix, *(mergeACLEntryCapabilities(&existing, pc)))
 			}
 		}
 	}
 	return a, nil
 }
 
+func mergeACLEntryCapabilities(existing *aclEntry, pc *PathCapabilities) (ret *aclEntry) {
+	ret = &aclEntry{}
+
+	// Ensure that we start out with a representative map based on existing
+	switch {
+	case existing.capabilitiesToMFAMap == nil && pc.MFAMethods == nil:
+		// Neither are using MFA so simply merge and return
+		ret.capabilitiesBitmap = existing.capabilitiesBitmap | pc.CapabilitiesBitmap
+		return ret
+
+	case existing.capabilitiesToMFAMap == nil && pc.MFAMethods != nil:
+		// Convert to a map so that we can merge in the methods
+		ret.capabilitiesToMFAMap = convertCapBitmapToMFAMap(existing.capabilitiesBitmap, nil)
+
+	case existing.capabilitiesToMFAMap != nil && pc.MFAMethods == nil:
+		// Start out using the existing map
+		ret.capabilitiesToMFAMap = existing.capabilitiesToMFAMap
+	}
+
+	// Now, create the map for the new path capabilities object
+	pathCapMFAMap := convertCapBitmapToMFAMap(pc.CapabilitiesBitmap, pc.MFAMethods)
+
+	/*
+		defer func() {
+			panic(fmt.Sprintf("existing:\n%#v\n, pc:\n%#v\n, currRetMap:\n%#v\n, pathCapMFAMap: \n%#v\n, ret:\n%#v\n", *existing, *pc, ret.capabilitiesToMFAMap, pathCapMFAMap, *ret))
+		}()
+	*/
+
+	switch {
+	case ret.capabilitiesToMFAMap == nil && pathCapMFAMap == nil:
+		// This really shouldn't happen, because in the first switch if the
+		// existing maps were nil it should have merged bitmaps and
+		// returned, but safety.
+		return ret
+
+	case pathCapMFAMap == nil:
+		// We have nothing(?) from the new pathcapabilities so return what we do have
+		return ret
+
+	case ret.capabilitiesToMFAMap == nil:
+		// In this case we had no existing capabilities, so use whatever has
+		// come down the line
+		ret.capabilitiesToMFAMap = pathCapMFAMap
+		return ret
+	}
+
+	// If we are at this point, both MFA maps are not nil, so we merge them.
+	for cap, mfas := range pathCapMFAMap {
+		ret.capabilitiesToMFAMap[cap] = append(ret.capabilitiesToMFAMap[cap], mfas...)
+	}
+
+	return ret
+}
+
+func convertCapBitmapToMFAMap(bitmap uint32, mfaMethods []string) map[uint32][]string {
```

Comment:
I think if we store PathCapabilities directly, this is unnecessary, since its just doing a look aside for the `mfaMethods` in every case